src/charm.py

```diff
@@ -0,0 +1,507 @@
+        self.nginx.configure_pebble_layer(tls=self._is_tls_ready)
+
+    def _on_nginx_prometheus_exporter_pebble_ready(self, _) -> None:
+        self.nginx_prometheus_exporter.configure_pebble_layer()
+
+    def _on_ingress_ready(self, event: IngressPerAppReadyEvent):
+        """Log the obtained ingress address.
+
+        This event refreshes the PrometheusRemoteWriteProvider address.
+        """
+        logger.info("Ingress for app ready on '%s'", event.url)
+
+    def _on_ingress_revoked(self, _) -> None:
+        """Log the ingress address being revoked.
+
+        This event refreshes the PrometheusRemoteWriteProvider address.
+        """
+        logger.info("Ingress for app revoked")
+
+    ######################
+    # === PROPERTIES === #
+    ######################
+
+    @property
+    def hostname(self) -> str:
+        """Unit's hostname."""
+        return socket.getfqdn()
+
+    @property
+    def _is_cert_available(self) -> bool:
+        return (
+            self.server_cert.enabled
+            and (self.server_cert.server_cert is not None)
+            and (self.server_cert.private_key is not None)
+            and (self.server_cert.ca_cert is not None)
+        )
+
+    @property
+    def _is_tls_ready(self) -> bool:
+        return (
+            self._nginx_container.can_connect()
+            and self._nginx_container.exists(CERT_PATH)
+            and self._nginx_container.exists(KEY_PATH)
+            and self._nginx_container.exists(CA_CERT_PATH)
+        )
+
+    @property
+    def loki_worker_relations(self) -> List[ops.Relation]:
+        """Returns the list of worker relations."""
+        return self.model.relations.get("loki_worker", [])
+
+    @property
+    def _workers_scrape_jobs(self) -> List[Dict[str, Any]]:
+        scrape_jobs = []
+        worker_topologies = self.cluster_provider.gather_topology()
+
+        for worker in worker_topologies:
+            job = {
+                "static_configs": [
+                    {
+                        "targets": [f"{worker['address']}:8080"],
+                    }
+                ],
+                # setting these as "labels" in the static config gets some of them
+                # replaced by the coordinator topology
+                # https://github.com/canonical/prometheus-k8s-operator/issues/571
+                "relabel_configs": [
+                    {"target_label": "juju_charm", "replacement": "loki-worker-k8s"},
+                    {"target_label": "juju_unit", "replacement": worker["unit"]},
+                    {"target_label": "juju_application", "replacement": worker["app"]},
+                    {"target_label": "juju_model", "replacement": self.model.name},
+                    {"target_label": "juju_model_uuid", "replacement": self.model.uuid},
+                ],
+            }
+            scrape_jobs.append(job)
+        return scrape_jobs
+
+    @property
+    def _nginx_scrape_jobs(self) -> List[Dict[str, Any]]:
```

Comment:
Would it make sense to move out some nginx related logic into a workload manager file? (We'd name the file the same as the container name in metadata.)

Reply:
We already have a `nginx.py`, you mean moving it there?